src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -0,0 +1,171 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "tcp_sockcm_ep.h"
+#include <ucs/sys/sock.h>
+#include <ucs/async/async.h>
+#include <ucs/arch/bitops.h>
+#include <ucs/sys/string.h>
+
+
+static UCS_F_ALWAYS_INLINE
+uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
+{
+    /* return the tcp sockcm connection manager this ep is using */
+    return ucs_container_of(cep->super.super.iface, uct_tcp_sockcm_t, super.iface);
+}
+
+ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_server_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(cep);
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr *server_addr;
+    ucs_async_context_t *async_ctx;
+    uct_tcp_sa_arg_t *sa_arg_ctx;
+    ucs_status_t status;
+    int events;
+
+    cep->state                   |= UCT_TCP_SOCKCM_EP_ON_CLIENT |
+                                    UCT_TCP_SOCKCM_EP_INIT;
+    cep->wireup.client.connect_cb = params->sockaddr_connect_cb.client;
+
+    server_addr = params->sockaddr->addr;
+    status = ucs_socket_create(server_addr->sa_family, SOCK_STREAM, &cep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the fd to non-blocking mode. (so that connect() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(cep->fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    /* try to connect to the server */
+    status = ucs_socket_connect(cep->fd, server_addr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        goto err_close_socket;
+    }
+    ucs_assert ((status == UCS_OK) || (status == UCS_INPROGRESS));
+
+    if (status == UCS_OK) {
+        cep->state &= ~UCT_TCP_SOCKCM_EP_INIT;
+        cep->state |= UCT_TCP_SOCKCM_EP_CONNECTED;
+        /* TODO: start sending the user's data */
+
+        goto out_print;
+    }
+
+    sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "client sa_arg_ctx");
+    if (sa_arg_ctx == NULL) {
+        ucs_error("failed to allocate memory for a client_ctx");
+        status = UCS_ERR_NO_MEMORY;
+        goto err_close_socket;
+    }
+
+    sa_arg_ctx->fd = cep->fd;
+    sa_arg_ctx->ep = cep;
+
+    /* Adding the arg to a list on the cm for cleanup purposes */
+    ucs_list_add_tail(&tcp_sockcm->sa_arg_list, &sa_arg_ctx->list);
+
+    events    = UCS_EVENT_SET_EVWRITE;    /* wait until connect() completes */
```

Comment:
why not provide it directly to `ucs_async_set_event_handler`?

Reply:
you are right, i will set it directly. it's just that in the future we will have several options for 'events' and it would be set before so will need the param. but for now, of course, can use it directly.

Reply:
> you are right, i will set it directly. it's just that in the future we will have several options for 'events' and it would be set before so will need the param. but for now, of course, can use it directly.

ok, thank you for the explanation